hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion6Strategy.java

```diff
@@ -70,47 +94,31 @@ public int getGeneratedVersion() {
 	}
 
 	@Override
-	public UUID generateUUID(SharedSessionContractImplementor session) {
+	public UUID generateUUID(final SharedSessionContractImplementor session) {
 		return generateUuid( session );
 	}
 
 	@Override
-	public UUID generateUuid(SharedSessionContractImplementor session) {
-		final long currentTimestamp = getCurrentTimestamp();
+	public UUID generateUuid(final SharedSessionContractImplementor session) {
+		final State state = lastState.updateAndGet( State::getNextState );
 
 		return new UUID(
 				// MSB bits 0-47 - most significant 32 bits of the 60-bit starting timestamp
-				currentTimestamp << 4 & 0xFFFF_FFFF_FFFF_0000L
+				state.lastTimestamp << 4 & 0xFFFF_FFFF_FFFF_0000L
 				// MSB bits 48-51 - version = 6
 				| 0x6000L
 				// MSB bits 52-63 - least significant 12 bits from the 60-bit starting timestamp
-				| currentTimestamp & 0x0FFFL,
+				| state.lastTimestamp & 0x0FFFL,
 				// LSB bits 0-1 - variant = 4
 				0x8000_0000_0000_0000L
 				// LSB bits 2-15 - clock sequence
-				| ( getSequence( currentTimestamp ) & 0x3FFFL ) << 48
-				// LSB bits 16-63 - pseudorandom data
-				| Holder.numberGenerator.nextLong() & 0xFFFF_FFFF_FFFFL
+				| ( state.lastSequence & 0x3FFFL ) << 48
```

Comment:
Current implementation already ensures that the value is in range, so no need for masking.

```suggestion
				| state.lastSequence << 48
```